accord-core/src/main/java/accord/topology/TopologyManager.java

```diff
@@ -0,0 +1,325 @@
+                EpochState currentEpoch = epochs[0];
+                pendingSync.remove(0).forEach(id -> currentEpoch.recordSyncComplete(id, prevSynced));
+            }
+            System.arraycopy(epochs, 0, nextEpochs, 1, epochs.length);
+
+            boolean prevSynced = epochs.length == 0 || epochs[0].syncComplete();
+            EpochState nextEpochState = new EpochState(topology, prevSynced);
+            nextEpochs[0] = nextEpochState;
+            return new Epochs(nextEpochs, pendingSync);
+        }
+
+        public void syncComplete(Node.Id node, long epoch)
+        {
+            Preconditions.checkArgument(epoch > 0);
+            if (epoch > maxEpoch - 1)
+            {
+                int idx = (int) (epoch - maxEpoch);
+                for (int i=pendingSyncComplete.size(); i<=idx; i++)
+                    pendingSyncComplete.add(new HashSet<>());
+
+                pendingSyncComplete.get(idx).add(node);
+            }
+            else
+            {
+                boolean prevSynced = epoch == minEpoch || get(epoch - 1).syncComplete();
+                boolean hasChanged = get(epoch).recordSyncComplete(node, prevSynced);
+                for (epoch++ ;hasChanged && epoch <= maxEpoch; epoch++)
+                {
+                    prevSynced = get(epoch - 1).syncComplete();
+                    hasChanged = get(epoch).updateState(prevSynced);
+                }
+            }
+        }
+
+        private EpochState get(long epoch)
+        {
+            if (epoch > maxEpoch || epoch < maxEpoch - epochs.length)
+                return null;
+
+            return epochs[(int) (maxEpoch - epoch)];
+        }
+
+        long canProcess(TxnRequestScope scope)
+        {
+            EpochState lastState = null;
+            for (int i=0, mi=scope.size(); i<mi; i++)
+            {
+                TxnRequestScope.EpochRanges requestRanges = scope.get(i);
+                EpochState epochState = get(requestRanges.epoch);
+
+                if (epochState != null)
+                {
+                    lastState = epochState;
+                }
+                else if (lastState != null && lastState.local.ranges().intersects(requestRanges.keys))
+                {
+                    // we don't have the most recent epoch, but still replicate the requested ranges
+                    continue;
+                }
+                else
+                {
+                    // we don't have the most recent epoch, and we don't replicate the requested ranges
+                    return scope.maxEpoch();
+                }
+
+                // validate requested ranges
+                KeyRanges localRanges = epochState.local.ranges();
+                if (!localRanges.intersects(requestRanges.keys))
+                    throw new RuntimeException("Received request for ranges not replicated by this node");
+            }
+            if (scope.maxEpoch() > 0)
+                missingEpochNotify.accept(scope.maxEpoch());
```

Comment:
Are we invoking this unnecessarily?

Reply:
no, but the variable name and documentation on the method it is meant to ultimately call are unclear

Reply:
Thanks. The confusion for me was that we also seemed to invoke this in `Node.receive` if we _didn't_ know an epoch. I can see now that this method is intended to be used differently by implementations, but it seems to also perform the same fetching by default. Are we relying on this behaviour? If not, might it be better (particularly from a test perspective) to have this no-op by default?